Warn if unsaved changes exist in assessment.

cahrens opened the pull request:
## [TNL-3870](https://openedx.atlassian.net/browse/TNL-3870)

This adds a warning if you try to browse away from a page where you have unsubmitted assessments. It covers peer assessments, staff override assessments, and staff full grade assessments (note you can have as many as 3 unsaved at once!). It builds on top of the infrastructure that warned if browsing away with an unsubmitted response.

To prevent the submission of a full grade assessment from wiping out a staff override assessment with unsaved changes, I had to do some refactoring of the full staff form. Because of this, now the "ungraded" and "checked out" counts will update when you check out an assessment for grading (which seems nice to me!).

### Sandbox
- [x] https://cahrens.sandbox.edx.org/courses/course-v1:ora+ora+ora/courseware/a4dfec19cf9b4a6fb5b18be6ccd9cecc/

### Testing
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @andy-armstrong 
- [x] Code review: @efischer19 
- [x] UI strings/error msgs review: @catong 
- [x] Product review: @explorerleslie

FYI: @scottrish 

### Post-review
- [x] Squash commits

cahrens commented:
Initial PR review was in #825.

I still need to respond to @catong's feedback in this PR.

andy-armstrong commented:
I just noticed that you don't catch changes made in self assessments.

andy-armstrong commented on openassessment/templates/openassessmentblock/staff_area/oa_staff_grade_learners_count.html:
```diff
@@ -0,0 +1,11 @@
+            {% endblocktrans %}
+        </span>
+    </span>
+</span>
```
Nit: missing newline

cahrens replied:
Fixed.

andy-armstrong commented on openassessment/xblock/static/js/src/lms/oa_base.js:
```diff
@@ -148,6 +158,40 @@ OpenAssessment.BaseView.prototype = {
+        }
+        else {
+            if (enabled) {
+                OpenAssessment.unsavedChanges[key] = message;
```
I wonder if you should track this per usage key, in the eventuality that we support multiple ORA problems on a single page. If you don't, then finishing one peer problem would mark them all as not dirty, potentially causing the loss of other changes.

andy-armstrong replied:
Another option might be to store the dirty state as data attributes on the DOM element in question. Then we can scope the state to each problem.

cahrens replied:
I decided to store them per usage ID. Storing on the DOM seemed attractive in some ways, but I couldn't think of a good place to store the error messages (since they really belong with the individual views).

andy-armstrong commented on openassessment/xblock/static/js/spec/lms/oa_peer.js:
```diff
@@ -115,4 +119,34 @@ describe("OpenAssessment.PeerView", function() {
+
+        expect(view.baseView.unsavedWarningEnabled()).toBe(true);
+
+        // When submitPeerAssessment is exected, the views will all re-render. However,
```
Typo: exected

andy-armstrong commented on openassessment/xblock/static/js/spec/lms/oa_staff_area.js:
```diff
@@ -557,18 +588,75 @@ describe('OpenAssessment.StaffAreaView', function() {
+            expect(staffArea.baseView.unsavedWarningEnabled()).toBe(false);
+        });
+
+        it('tracks unsubmitted assessments in multiple views', function() {
```
Nit: I'd move this test out of the 'Grade Available Responses' section since it tests cross-view capabilities.

cahrens replied:
I moved this test point out.

andy-armstrong commented:
:+1: Very comprehensive, both in functionality and tests! 

If you are concerned about supporting multiple active problems on a page you could address this now, or we could just add it as acceptance criteria to the ticket for solving the whole issue (https://openedx.atlassian.net/browse/TNL-1593)

cahrens commented:
@catong and @andy-armstrong -- I updated doc strings and added warnings for the self assessment workflow. The sandbox has been updated.

catong commented:
@cahrens Looks great, thanks! :+1:

andy-armstrong commented:
:+1:

cahrens commented:
@andy-armstrong I addressed your feedback in https://github.com/edx/edx-ora2/commit/bbd787cc72eb0080234856d233a1d05115dedd25 (if you want to take a look after the holidays). I'm going to squash this PR now to merge to ora-staff-grading.
